Initial findfeatures app conversion + gtests

AustinSanders opened the pull request:
Converted findfeatures app to a function and added gtests

## Description
Converted findfeatures/main.cpp into findfeatures/findfeatures.cpp and findfeatures/findfeatures.h.  Added 8 new gtests.

## Related Issue
Test conversion

## Motivation and Context
Frees 576Mb testing space

## How Has This Been Tested?
8 Gtests.  Currently rerunning the entire suite of tests since we altered some fixtures.

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation change (update to the documentation; no code change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- - [ ] My code follows the code style of this project. -->
- [x] I have read and agree to abide by the [Code of Conduct](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/Code-Of-Conduct.md)
- [x] I have read the [**CONTRIBUTING**](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] I have added myself to the [.zenodo.json](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/.zenodo.json) document.
- [ ] I have added any user impacting changes to the [CHANGELOG.md](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CHANGELOG.md) document.

## Licensing
This project is mostly composed of free and unencumbered software released into the public domain, and we are unlikely to accept contributions that are not also released into the public domain. Somewhere near the top of each file should have these words:

> This work is free and unencumbered software released into the public domain. In jurisdictions that recognize copyright laws, the author or authors of this software dedicate any and all copyright interest in the software to the public domain.

- [x] I dedicate any and all copyright interest in this software to the public domain. I make this dedication for the benefit of the public at large and to the detriment of my heirs and successors. I intend this dedication to be an overt act of relinquishment in perpetuity of all present and future rights to this software under copyright law.

AustinSanders committed:
Initial findfeatures conversion + gtests

AustinSanders committed:
Delete makefiles

acpaquette commented on isis/src/control/apps/findfeatures/findfeatures.h:
```diff
@@ -0,0 +1,12 @@
+#ifndef findfeatures_h // Change this to your app name in all lower case suffixed with _h (e.g. campt_h, cam2map_h etc.)
+#define findfeatures_h
+
+#include "Cube.h"
```
Shouldn't need this extra cube include here.

acpaquette commented on isis/tests/Fixtures.cpp:
```diff
@@ -312,9 +313,44 @@ namespace Isis {
+      cube3->write(line3);
+    }
+
+    cube1->reopen("rw");
+    cube2->reopen("rw");
+    cube3->reopen("rw");
```
With the addition of these reopens it may be worth while to remove the reopens for cube1 and cube2 on lines 299 and 311

acpaquette commented on isis/tests/FunctionalTestsFindFeatures.cpp:
```diff
@@ -0,0 +1,215 @@
+  ControlNet network(options.GetFileName("ONET"));
+  ControlPoint *pt = network.GetPoint("test_network_0001");
+  SurfacePoint sp = pt->GetAdjustedSurfacePoint();
+  Latitude lat = sp.GetLatitude();
+  Longitude lon = sp.GetLongitude();
+  ASSERT_EQ(lat.toString(), "");
+  ASSERT_EQ(lon.toString(), "");
```
Kinda weird that these are blank but we're kinda assuming it's in a working state

AustinSanders replied:
The input cubes (fixtures) didn't have lat/lon information, and since we're getting geometries from those input cubes, it left them blank.  Not sure if (or how) to add geometries to the fixtures.

kaitlyndlee replied:
I think you would need a cube with a Mapping group, but I could be completely wrong. The projTestCube in the defaultCube fixture has one.

acpaquette commented on isis/src/control/apps/findfeatures/main.cpp:
```diff
@@ -1,377 +1,34 @@
   }
 
-  return;
+  SessionLog::TheLog().AddResults(results);
```
Out of curiosity where was this in the old code/what did it replace?

AustinSanders replied:
The app used both logs and guilogs, so I tried to create a mix of the two templates.  This line comes from the [gui log section](https://github.com/USGS-Astrogeology/ISIS3/wiki/Writing-ISIS3-Tests-Using-Gtest-and-Ctest#application-with-guilogs).

acpaquette reviewed (commented):
Overall looks good! Couple small things to address.

AustinSanders committed:
Addressed PR feedback

Removed unnecessary cube.h from findfeatures.h
Removed additional cube reopens from threeimagenetwork fixture.
Added a comment clarifying empty lat/lon values in match test.

kaitlyndlee commented on isis/tests/Fixtures.cpp:
```diff
@@ -307,14 +307,48 @@ namespace Isis {
+    int pixelValue = 1;
+    for(line.begin(); !line.end(); line++) {
+      for(int i = 0; i < line.size(); i++) {
+        line[i] = (double) (pixelValue %255);
```
Since the three cube's pixel type is real, we don't need to make sure the pixel values are between 0-255. We can keep this though if we prefer to restrict the values.

kaitlyndlee commented on isis/tests/FunctionalTestsFindFeatures.cpp:
```diff
@@ -0,0 +1,216 @@
+                           "description=new",
+                           "debug=false"};
+  UserInterface options(APP_XML, args);
+  Pvl log;
```
Since we don't use the log data in the test and it can be null in the findfeatures call, can we remove this variable?

AustinSanders committed:
Removed template comment

AustinSanders committed:
Removed unused AddFeatures function

AustinSanders committed:
Edited test for lat/lon in geomsource

acpaquette reviewed (dismissed):
Looks good! :+1:

AustinSanders commented:
@acpaquette @kaitlyndlee  @Kelvinrr It looks like the MaxPoints test is failing on Mac, and that it is expecting 19 points rather than the 22 points found on other OSs + my own Mac machine.  I'm not sure how to resolve this, particularly due to the nondeterministic nature of the FindFeatures algorithm(s).  Any ideas?

AustinSanders committed:
Edited maxpoints test to compare multiple runs of findfeatures

acpaquette commented:
@AustinSanders It looks like the solution you applied worked. I think it's a good way to go about it. The only other way I could think to try is have the code see if it's on mac or linux then check for the expect operating system value. I think what you did is much better though!
